### ISSUE TITLE
Test what happens if we minify views

### ISSUE THREAD

**Issue by garethbowen** [medic/medic-webapp#4082]
Once upon a time we tested view performance and found that views with a lot of source took longer to index than views with not much source, even if the logic returned at the same point. It's possible that removing comments, using short variable names, and other minification tricks will improve build performance.

Investigate whether view minification delivers any performance benefits.

**Comment by garethbowen** [comment]
I tested on a CDB1 database with 40000 docs by creating a temporary view based on one of our actual views, then I minified it with Uglify v3, then I removed whitespace only. Then I measured the time it took to get a result in seconds. To break the cache I made a minimal change to the view each round.

Results:

| View | Test 1 | Test 2 | Test 3 | Test 4 | Mean |
|--|--|--|--|--|--|
| Original | 35.106 | 37.210 | 31.338 | 32.126 | 33.945 |
| Minified | 50.651 | 50.563 | 48.139 | 50.748 | 50.02525 |
| Whitespace removed | 33.308 | 31.293 | 35.770 | 31.599 | 32.9925 |

It looks like the minification breaks some logic somewhere to make it so much slower. The whitespace removal is within the margin of error of the original so there's no statistically significant improvement.

Conclusion: we shouldn't bother with any view minification.

**Comment by garethbowen** [comment]
Reopening for discussion.

**Comment by garethbowen** [comment]
Adding my code samples in case we want to run more tests...

### Original

```js
function(doc) {

  var getLineage = function(contact2) {
    var lineage = [];
    while (contact2) {
      if (contact2._id) {
        lineage.push(contact2._id);
      }
      contact2 = contact2.parent;
    }
    return lineage;
  };

  var emitMessage = function(doc, contact, phone, message) {
    var id = (contact && contact._id) || phone;
    emit([ id, doc.reported_date ], {
      id: doc._id,
      date: doc.reported_date,
      message: message,
      contact: contact && contact._id,
      lineage: getLineage(contact && contact.parent)
    });
  };

  if (doc.type === 'data_record' && !doc.form) {
    if (doc.kujua_message) {
      // outgoing
      doc.tasks.forEach(function(task) {
        var message = task.messages[0];
        emitMessage(doc, message.contact, message.to, message.message);
      });
    } else if (doc.sms_message) {
      // incoming
      emitMessage(doc, doc.contact, doc.from, doc.sms_message.message);
    }
  }
}
```

### Uglified

```js
function(z) {
var l=function(e){for(var a=[];e;)e._id&&a.push(e._id),e=e.parent;return a},m=function(e,a,d,s){var t=a&&a._id||d;emit([t,e.reported_date],{id:e._id,date:e.reported_date,message:s,contact:a&&a._id,lineage:l(a&&a.parent)})};"data_record"!==z.type||z.form||(z.kujua_message?z.tasks.forEach(function(e){var a=e.messages[0];m(z,a.contact,a.to,a.message)}):z.sms_message&&m(z,z.contact,z.from,z.sms_message.message));
}
```

### Whitespace removed

```js
function(doc){
var getLineage=function(contact2){
var lineage=[];
while(contact2){
if(contact2._id){lineage.push(contact2._id);}
contact2=contact2.parent;
}return lineage;};
var emitMessage=function(doc,contact,phone,message){
var id=(contact && contact._id) || phone;
emit([id,doc.reported_date ],{id:doc._id,date:doc.reported_date,message:message,contact:contact && contact._id,lineage:getLineage(contact && contact.parent)});
};
if(doc.type === 'data_record' && !doc.form){
if(doc.kujua_message){
doc.tasks.forEach(function(task){
var message=task.messages[0];
emitMessage(doc,message.contact,message.to,message.message);
});
} else if(doc.sms_message){emitMessage(doc,doc.contact,doc.from,doc.sms_message.message);}
}}
```

**Comment by garethbowen** [comment]
Testing on a large database (4.5m docs) yielded much more significant results from 220 minutes down to 142 minutes or 35% faster for the minified version.

#### Questions

1. Does this also affect CouchDB 2 or has it been fixed?
2. I tested on temporary views - are they the same as real views?
3. Does this also affect PouchDB?

#### Next steps

Upgrade the large instance to CouchDB 2 and create the views attached to a ddoc to test questions 1 and 2 above.

**Comment by alxndrsn** [comment]
To me this is both good enough and simple enough to warrant immediate implementation.  Although I guess it will mean view regeneration for projects which upgrade, even if views haven't changed 😆

**Comment by garethbowen** [comment]
I tested on PouchDB...

| View | Test 1 | Test 2 | Test 3 | Mean | Relative time | Size | Relative size |
|--|--|--|--|--|--|--|--|
| Original | 24.865 | 26.607 | 25.940 | 25.804 | 100% | 1345 chars | 100% |
| Minified | 26.963 | 27.702 | 27.051 | 27.238 | 105% | 711 chars | 53% |

So PouchDB doesn't perform any faster and potentially performs slower, but it would save us about 9KB of medic-client ddoc size.

Because the medic-client views are executed in couchdb as well as pouchdb it's probably worth doing them too.

I still need to test CouchDB v2 to see if the performance issue has been fixed.

**Comment by garethbowen** [comment]
CouchDB 2 on alpha.dev with around 2000 docs.

| View | Test 1 | Test 2 | Test 3 | Test 4 | Mean | Relative |
|--|--|--|--|--|--|--|
| Original | 1342 | 1417 | 1356 | 1648 | 1441 | 100%  |
| Minified | 1635 | 1595 | 1310 | 1411 | 1488 | 103%  |

Either this issue is fixed in 2.0 _or_ there were too few docs to expose the issue. I'll try to find a server with more docs before closing this off.

**Comment by garethbowen** [comment]
I've tested this on a db with 8.3m docs.

| View | Test 1 |
|--|--|--|
| Original | 41 minutes |
| Minified | 40 minutes |

So it's now basically the same and I'm going to close this out and assert that it makes no difference.

It's worth noting the difference between CouchDB 1 and 2 where the minified version came down from 142 minutes for 4.5m docs to 40 minutes for 8.3m docs, so it appears that view indexing is much faster in v2.
